### DISCUSSION TITLE
Perf advice for points-in-polygon test

### DISCUSSION THREAD

**Discussion by hyperknot** [shapely/shapely#1351] [Q&A]
I've read through the docs, the source code but before I start implementing my function I'd like to ask for advice here.

I have list of 2D points (lat, lon) and a single 2D polygon (possibly with holes). What is the fastest algorithm which checks if at least one point is within the polygon? 

I'd like to call this function multiple times, with multiple point-sets, but always with the same polygon. 

I'm thinking about something like:
1. Load and prepare polygon geom at startup
2. Use `_construct_points` on the point list. I'd need to unzip them manually first, maybe there is a faster numpy algorithm which unzips in a single pass? https://github.com/shapely/shapely/blob/f048c1816ad5782c2e70b42c55e5d74fd19ce7e7/shapely/vectorized/__init__.py#L9 
3. Now what should I call on this? `pre_geom.intersects(np_points)`? or `shapely.intersects(prep_geom, np_points)`?

I couldn't really find documentation about vectorised and also why only 2 methods are supported there. It seems like a very simple wrapper over numpy-inside-points and prepared geometry binary operations. Is there something I'm missing?

**Comment by jorisvandenbossche** [comment]
@hyperknot what you outline sounds as a good approach. For a specific case of one polygon and many points, using a prepared polygon will probably give the best results.

Note that the main development branch of Shapely (what will get released as Shapely 2.0) contains vectorized operations for almost all functions and methods (and not only the small subset which exists in the `vectorized` module you point to).

Using the main branch, this would look like:

```
>>> import shapely
>>> polygon = shapely.box(0, 0, 1, 1)
>>> points = shapely.points(np.random.randn(10000, 2))
>>> shapely.prepare(polygon)
>>> shapely.contains(polygon, points)
array([False, False, False, ...,  True, False,  True])
```

(and then the contains function will automatically see that the polygon was prepared)

The above won't work yet for the released version of Shapely. But for that you can check the example at https://shapely.readthedocs.io/en/stable/manual.html#prepared-geometry-operations. But note that the prepared geometries in the released version of shapely won't work with the functions of the `vectorized` module.

For a released version of the code above, you can check `pygeos` (https://pygeos.readthedocs.io/en/stable/)

**Reply by hyperknot** [reply]
Thanks! Are there binary wheels available for the 2.0 branch beta?

**Reply by jorisvandenbossche** [reply]
Not yet, I hope we get to that soon, but for now you will need to install from source, or use the pygeos package (if you replace "shapely" with "pygeos" in the example above, it works exactly the same)

**Reply by hyperknot** [reply]
Great, I'll use pygeos then!
